integration-tests/deployment/keystone/update_nodes_test.go

```diff
@@ -0,0 +1,516 @@
+			want: &kslib.UpdateNodesResponse{
+				NodeParams: []kcr.CapabilitiesRegistryNodeParams{
+					{
+						NodeOperatorId:      1,
+						P2pId:               p2pId("peerID_1"),
+						HashedCapabilityIds: nil, // checked dynamically based on the request
+						Signer:              [32]byte{0: 1, 1: 2},
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "one node, two capabilities",
+			args: args{
+				lggr: lggr,
+				req: &kslib.UpdateNodesRequest{
+					P2pToCapabilities: map[p2pkey.PeerID][]kcr.CapabilitiesRegistryCapability{
+						p2pId("peerID_1"): []kcr.CapabilitiesRegistryCapability{
+							{
+								LabelledName:   "cap1",
+								Version:        "1.0.0",
+								CapabilityType: 0,
+							},
+							{
+								LabelledName:   "cap2",
+								Version:        "1.0.1",
+								CapabilityType: 2,
+							},
+						},
+					},
+					NopToNodes: map[kcr.CapabilitiesRegistryNodeOperator][]*kslib.P2PSigner{
+						testNop(t, "nop1"): []*kslib.P2PSigner{
+							{
+								P2PKey: p2pId("peerID_1"),
+								Signer: [32]byte{0: 1, 1: 2},
+							},
+						},
+					},
+					Chain:    chain,
+					Registry: nil, // set in test to ensure no conflicts
+				},
+			},
+			want: &kslib.UpdateNodesResponse{
+				NodeParams: []kcr.CapabilitiesRegistryNodeParams{
+					{
+						NodeOperatorId:      1,
+						P2pId:               p2pId("peerID_1"),
+						HashedCapabilityIds: nil, // checked dynamically based on the request
+						Signer:              [32]byte{0: 1, 1: 2},
+					},
+					{
+						NodeOperatorId:      1,
+						P2pId:               p2pId("peerID_1"),
+						HashedCapabilityIds: nil, // checked dynamically based on the request
+						Signer:              [32]byte{0: 1, 1: 2},
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "twos node, one shared capability",
+			args: args{
+				lggr: lggr,
+				req: &kslib.UpdateNodesRequest{
+					P2pToCapabilities: map[p2pkey.PeerID][]kcr.CapabilitiesRegistryCapability{
+						p2pId("peerID_1"): []kcr.CapabilitiesRegistryCapability{
+							{
+								LabelledName:   "cap1",
+								Version:        "1.0.0",
+								CapabilityType: 0,
+							},
+						},
+						p2pId("peerID_2"): []kcr.CapabilitiesRegistryCapability{
+							{
+								LabelledName:   "cap1",
+								Version:        "1.0.0",
+								CapabilityType: 0,
+							},
+						},
+					},
+					NopToNodes: map[kcr.CapabilitiesRegistryNodeOperator][]*kslib.P2PSigner{
+						testNop(t, "nopA"): []*kslib.P2PSigner{
+							{
+								P2PKey: p2pId("peerID_1"),
+								Signer: [32]byte{0: 1, 31: 1},
+							},
+						},
+						testNop(t, "nopB"): []*kslib.P2PSigner{
+							{
+								P2PKey: p2pId("peerID_2"),
+								Signer: [32]byte{0: 2, 31: 2},
+							},
+						},
+					},
+					Chain:    chain,
+					Registry: nil, // set in test to ensure no conflicts
+				},
+			},
+			want: &kslib.UpdateNodesResponse{
+				NodeParams: []kcr.CapabilitiesRegistryNodeParams{
+					{
+						NodeOperatorId:      1,
```

Comment:
/nit `ID` vs `Id` for our linting rules

Reply:
generated struct field names from solidity